src/cargo/util/rustc.rs

```diff
@@ -38,14 +38,31 @@ impl Rustc {
     /// If successful this function returns a description of the compiler along
     /// with a list of its capabilities.
     pub fn new(
-        path: PathBuf,
+        mut path: PathBuf,
         wrapper: Option<PathBuf>,
         workspace_wrapper: Option<PathBuf>,
         rustup_rustc: &Path,
         cache_location: Option<PathBuf>,
     ) -> CargoResult<Rustc> {
         let _p = profile::start("Rustc::new");
 
+        // In order to avoid calling through rustup multiple times, we first ask
+        // rustc to give us the "resolved" rustc path, and use that instead. If
+        // this doesn't give us a path, then we just use the original path such
+        // that the following logic can handle any resulting errors normally.
+        let mut cmd = ProcessBuilder::new(&path);
+        cmd.arg("--print=rustc-path");
+        if let Ok(output) = cmd.output() {
```

Comment:
@CAD97 have you benchmarked this to ensure this gives the expected gains?

> This increases the amount of rustc invocations before building anything from 3 to 4:

Do we have a good case to benchmark to see the impact of bumping the rustc invocations pre-build from 3-4?

Reply:
Take the below with a healthy helping of skepticism that the results are meaningful.

<details><summary>Ad-hoc measurement</summary>

For maximum impact, checking an empty bin with no dependencies, averaging over 128 iterations:

```nushell
D:\git\cad97\playground〉ls src | select name size
╭───┬─────────────┬──────╮
│ # │    name     │ size │
├───┼─────────────┼──────┤
│ 0 │ src\main.rs │ 13 B │
╰───┴─────────────┴──────╯
D:\git\cad97\playground〉open .\Cargo.toml
╭──────────────┬───────────────────╮
│ dependencies │ {record 0 fields} │
│ package      │ {record 4 fields} │
╰──────────────┴───────────────────╯
D:\git\cad97\playground〉open .\bench.nu
rustup override set stage1

let iter_count = 128

for _ in 1..$iter_count { cargo clean; benchmark { cargo check -q } } | math avg
with-env { RUSTC: $"(rustup which rustc | str trim)" } {
    for _ in 1..$iter_count { cargo clean; benchmark { cargo check -q } } | math avg
}
for _ in 1..$iter_count { cargo clean; benchmark { ./cargo.exe check -q } } | math avg

rustup override remove
D:\git\cad97\playground〉source bench.nu
info: override toolchain for 'D:\git\cad97\playground' set to 'stage1'
220ms 263µs 608ns
219ms 523µs 601ns
277ms 484µs 549ns
info: override toolchain for 'D:\git\cad97\playground' removed
```

| strategy | timing | delta |
| -: | -: | -: |
| existing | `0.220263608`s | 0 |
| $RUSTC | `0.219523601`s | `-0.000740007`s |
| --print | `0.277484549`s | `+0.057220941`s |

-----
</details>

It seems at least on Windows an extra rustc startup invocation is about a 25% overhead for an absolutely empty crate, which seems about right, seeing as we're going from 4 to 5 invocations. (3 to 4 during startup, plus the one to actually build the crate.)

<details><summary>Scaling up to 27 empty binary targets...</summary>

```nushell
D:\git\cad97\playground〉ls bin | length
26
D:\git\cad97\playground〉source bench.nu
info: override toolchain for 'D:\git\cad97\playground' set to 'stage1'
343ms 61µs 896ns
360ms 891µs 528ns
421ms 625µs 731ns
info: override toolchain for 'D:\git\cad97\playground' removed
```

| strategy | timing | delta |
| -: | -: | -: |
| existing | `0.343061896`s | 0 |
| $RUSTC | `0.360891528`s | `-0.017829632`s |
| --print | `0.421625731`s | `+0.078563835`s |

-----
</details>

Interestingly, this sees the absolute overhead go up (~21ms, 37%) for some reason (noise? this patch normalizing rustc too late?) but the relative overhead fall (by about 3 percentage points). This needs more sophisticated benchmarks, and to make sure that the compared cargos are compiled by the same rustc (which I do not have).

This result either means my measurements are bunk, somethings wrong with the patch, or something weird is going on with the OS interactions to make the result nonintuitive.

Reply:
Slightly better comparison with locally compiled `cargo-main` and `cargo-patched` gives (average of 128 iterations)

| strategy | 27 binaries | delta | 1 binary | delta |
| -: | -: | -: | -: | -: |
| as-is | `0.424288469`s | - | `0.296789453`s | - |
| $RUSTC | `0.321916291`s | -`0.102372178`s | `0.183639414`s | -`0.113150039`s |
| --path | `0.414090124`s | -`0.010198345`s | `0.277666412`s | -`0.019123041`s |

... but these numbers make me feel the measurement still isn't measuring what I'm trying to measure (that being just the impact of this patch).